packages/transaction-messages/src/use-address-lookup-tables.ts

```diff
@@ -0,0 +1,86 @@
+// Look up the address in lookup tables, return a lookup meta if it is found in any of them
+function findAddressInLookupTables(
+    address: Address,
+    role: AccountRole.READONLY | AccountRole.WRITABLE,
+    addressesByLookupTableAddress: AddressesByLookupTableAddress,
+): IAccountLookupMeta | undefined {
+    for (const [lookupTableAddress, addresses] of Object.entries(addressesByLookupTableAddress)) {
+        for (let i = 0; i < addresses.length; i++) {
+            if (address === addresses[i]) {
+                return {
+                    address,
+                    addressIndex: i,
+                    lookupTableAddress: lookupTableAddress as Address,
+                    role,
+                };
+            }
+        }
+    }
+}
+
+type TransactionMessageNotLegacy = Exclude<TransactionMessage, { version: 'legacy' }>;
+
+export function useAddressLookupTablesForTransactionMessage<
+    TTransactionMessage extends TransactionMessageNotLegacy = TransactionMessageNotLegacy,
+>(
+    transactionMessage: TTransactionMessage,
+    addressesByLookupTableAddress: AddressesByLookupTableAddress,
+): TTransactionMessage {
+    // All signers will need to be static keys, so we shouldn't replace them if they're used elsewhere in the transaction,
+    // even if they're also in a lookup table
+    const signers = new Set(
+        transactionMessage.instructions
+            .flatMap(i => (i.accounts ?? []).filter(a => isSignerRole(a.role)))
+            .map(a => a.address),
+    );
+    if ('feePayer' in transactionMessage) {
+        signers.add((transactionMessage as ITransactionMessageWithFeePayer).feePayer);
+    }
+
+    const lookupTableAddresses = new Set(Object.values(addressesByLookupTableAddress).flatMap(a => a));
+
+    const newInstructions: IInstruction[] = [];
+    for (const instruction of transactionMessage.instructions) {
+        if (!instruction.accounts) {
+            newInstructions.push(instruction);
+            continue;
+        }
+
+        const newAccounts: Mutable<NonNullable<IInstruction['accounts']>> = [];
+        for (const account of instruction.accounts) {
+            // If the address is a signer (of any instruction), or already a lookup, or is not in any lookup tables, return as-is
+            if (
+                signers.has(account.address) ||
+                'lookupTableAddress' in account ||
+                !lookupTableAddresses.has(account.address) ||
+                isSignerRole(account.role)
+            ) {
+                newAccounts.push(account);
+                continue;
+            }
+
+            // We already checked it's in one of the lookup tables
+            const lookupMetaAccount = findAddressInLookupTables(
+                account.address,
+                account.role,
+                addressesByLookupTableAddress,
+            )!;
+            newAccounts.push(lookupMetaAccount);
+        }
+
+        newInstructions.push({ ...instruction, accounts: newAccounts });
```

Comment:
I think there's a bunch of `Object.freeze` that has to happen in here, and get tested like [this](https://github.com/solana-labs/solana-web3.js/blob/9239e6ec972b4de9f0d15b197fbef1d2871759d9/packages/signers/src/__tests__/add-signers-test.ts#L126-L141).